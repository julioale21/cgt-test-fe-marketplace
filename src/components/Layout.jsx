import { Outlet } from 'react-router-dom';
import { Header } from '../components/shared/Header';
import { Box } from '@mui/material';

export function Layout() {
  const cartItemsCount = 3;

  return (
    <Box sx={{ display: 'flex', flexDirection: 'column', minHeight: '100vh' }}>
      <Header cartItemsCount={cartItemsCount} />

      <Box component="main" sx={{ flexGrow: 1, p: 3 }}>
        <Outlet />
      </Box>

      <Box component="footer" sx={{ p: 2, textAlign: 'center', mt: 'auto' }}>
        <p>Â© 2025</p>
      </Box>
    </Box>
  );
}
