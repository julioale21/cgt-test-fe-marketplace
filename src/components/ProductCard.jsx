import React from 'react';

import { Link } from 'react-router-dom';

import { Button, Card, CardActions, CardContent, CardMedia, Grid, Typography } from '@mui/material';
import ShoppingCartIcon from '@mui/icons-material/ShoppingCart';

const ProductCard = ({ product }) => {
  return (
    <Grid item xs={12} sm={6} md={4} lg={3}>
      <Card sx={{ height: '100%', display: 'flex', flexDirection: 'column' }}>
        <CardMedia
          component="img"
          height="200"
          image={product.image}
          alt={product.name}
          sx={{
            objectFit: 'cover',
            height: '200px'
          }}
        />
        <CardContent sx={{ flexGrow: 1 }}>
          <Typography gutterBottom variant="h6" component="h2">
            {product.name}
          </Typography>
          <Typography variant="body2" color="text.secondary" sx={{ mb: 2 }}>
            {product.description}
          </Typography>
          <Typography variant="h6" color="primary" sx={{ mt: 2 }}>
            ${product.price.toFixed(2)}
          </Typography>
          <Typography variant="body2" color="text.secondary" sx={{ mt: 1 }}>
            Formats: {product.format.join(', ')}
          </Typography>
        </CardContent>
        <CardActions sx={{ justifyContent: 'space-between' }}>
          <Button component={Link} to={`/products/${product.id}`} size="small" color="primary">
            View Details
          </Button>
          <Button size="small" variant="contained" color="primary" startIcon={<ShoppingCartIcon />}>
            Add to Cart
          </Button>
        </CardActions>
      </Card>
    </Grid>
  );
};

export { ProductCard };
