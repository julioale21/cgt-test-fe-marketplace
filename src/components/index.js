export { ContainedButton } from "./buttons/ContainedButton";
export { Header } from "./shared/Header";
export { Layout } from "./Layout";
export { OutlinedButton } from "./buttons/OutlinedButton";
export { ProductList } from "./ProductList";
export {ProductCard} from "./ProductCard";
