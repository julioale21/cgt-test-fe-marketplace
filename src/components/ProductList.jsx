import React from 'react';
import { Grid } from '@mui/material';
import { ProductCard } from './ProductCard';

const ProductList = ({ products }) => {
  return (
    <Grid container spacing={4}>
      {products.map((product) => (
        <ProductCard key={product.id} product={product} />
      ))}
    </Grid>
  );
};

export { ProductList };
