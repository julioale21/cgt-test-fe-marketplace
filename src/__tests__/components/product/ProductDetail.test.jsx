import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ProductDetails } from '../../../components';

jest.mock('../../../components/product/ProductHeader', () => ({
  ProductHeader: ({ name, rating, category }) => (
    <div data-testid="product-header">
      {name} - {rating} - {category}
    </div>
  )
}));

jest.mock('../../../components/product/ProductSpecs', () => ({
  ProductSpecs: ({ format, polygons, textures }) => (
    <div data-testid="product-specs">
      {format.join(',')} - {polygons} - {textures}
    </div>
  )
}));

describe('ProductDetails', () => {
  const mockProduct = {
    name: 'Test Product',
    rating: 4.5,
    category: 'Test Category',
    price: 99.99,
    description: 'Test description',
    format: ['FBX', 'OBJ'],
    polygons: 'High',
    textures: '4K'
  };

  const mockOnAddToCart = jest.fn();

  const renderComponent = (props = {}) => {
    const defaultProps = {
      product: mockProduct,
      isInCart: false,
      onAddToCart: mockOnAddToCart,
      ...props
    };
    return render(<ProductDetails {...defaultProps} />);
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot when product is not in cart', () => {
    const { container } = render(
      <ProductDetails product={mockProduct} isInCart={false} onAddToCart={mockOnAddToCart} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when product is in cart', () => {
    const { container } = render(
      <ProductDetails product={mockProduct} isInCart={true} onAddToCart={mockOnAddToCart} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render all product information', () => {
    renderComponent();

    expect(screen.getByTestId('product-header')).toBeInTheDocument();
    expect(screen.getByText('$99.99')).toBeInTheDocument();
    expect(screen.getByText('Test description')).toBeInTheDocument();
    expect(screen.getByTestId('product-specs')).toBeInTheDocument();
    expect(screen.getByText('Cantidad:')).toBeInTheDocument();
    expect(screen.getByText('Add to Cart')).toBeInTheDocument();
  });

  it('should increment quantity when plus button is clicked', () => {
    renderComponent();
    const incrementButton = screen.getByTestId('AddIcon').parentElement;

    fireEvent.click(incrementButton);

    expect(screen.getByText('2')).toBeInTheDocument();
  });

  it('should decrement quantity when minus button is clicked', () => {
    renderComponent();
    const incrementButton = screen.getByTestId('AddIcon').parentElement;
    const decrementButton = screen.getByTestId('RemoveIcon').parentElement;
    fireEvent.click(incrementButton);

    fireEvent.click(decrementButton);

    expect(screen.getByText('1')).toBeInTheDocument();
  });

  it('should not decrement quantity below 1', () => {
    renderComponent();
    const decrementButton = screen.getByTestId('RemoveIcon').parentElement;

    fireEvent.click(decrementButton);

    expect(screen.getByText('1')).toBeInTheDocument();
  });

  it('should not increment quantity above 50', () => {
    renderComponent();
    const incrementButton = screen.getByTestId('AddIcon').parentElement;
    Array(51)
      .fill(null)
      .forEach(() => fireEvent.click(incrementButton));

    expect(screen.getByText('50')).toBeInTheDocument();
  });

  it('should call onAddToCart with selected quantity', () => {
    renderComponent();
    const incrementButton = screen.getByTestId('AddIcon').parentElement;
    const addToCartButton = screen.getByText('Add to Cart');
    fireEvent.click(incrementButton);
    fireEvent.click(incrementButton);

    fireEvent.click(addToCartButton);

    expect(mockOnAddToCart).toHaveBeenCalledWith(3);
  });

  it('should display different button state when product is in cart', () => {
    renderComponent({ isInCart: true });

    const button = screen.getByText('Already in Cart');

    expect(button).toBeInTheDocument();
    expect(button).toHaveClass('MuiButton-outlined');
  });

  it('should display maximum quantity message', () => {
    renderComponent();

    expect(screen.getByText('MÃ¡ximo 50 unidades')).toBeInTheDocument();
  });

  it('should maintain quantity state between renders', () => {
    const { rerender } = renderComponent();
    const incrementButton = screen.getByTestId('AddIcon').parentElement;
    fireEvent.click(incrementButton);

    rerender(
      <ProductDetails product={mockProduct} isInCart={false} onAddToCart={mockOnAddToCart} />
    );

    expect(screen.getByText('2')).toBeInTheDocument();
  });
});
