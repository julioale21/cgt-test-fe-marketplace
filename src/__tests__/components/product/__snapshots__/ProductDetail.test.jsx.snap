// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails should match snapshot when product is in cart 1`] = `
<div>
  <div
    class="MuiBox-root css-19midj6"
  >
    <div
      data-testid="product-header"
    >
      Test Product
       - 
      4.5
       - 
      Test Category
    </div>
    <h4
      class="MuiTypography-root MuiTypography-h4 css-1uuypgi-MuiTypography-root"
    >
      $
      99.99
    </h4>
    <p
      class="MuiTypography-root MuiTypography-body1 css-o5kk20-MuiTypography-root"
    >
      Test description
    </p>
    <div
      class="MuiBox-root css-1yjvs5a"
    >
      <div
        data-testid="product-specs"
      >
        FBX,OBJ
         - 
        High
         - 
        4K
      </div>
    </div>
    <div
      class="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <div
        class="MuiBox-root css-j0ozid"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-k16oju-MuiTypography-root"
        >
          Cantidad:
        </p>
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-schvjd-MuiPaper-root"
        >
          <button
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeSmall css-1374xo4-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            id=":r3:"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiIconButton-loadingIndicator css-165vu30-MuiIconButton-loadingIndicator"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="RemoveIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13H5v-2h14z"
              />
            </svg>
          </button>
          <p
            class="MuiTypography-root MuiTypography-body1 css-144jna6-MuiTypography-root"
          >
            1
          </p>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1374xo4-MuiButtonBase-root-MuiIconButton-root"
            id=":r4:"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-loadingIndicator css-165vu30-MuiIconButton-loadingIndicator"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="AddIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
              />
            </svg>
          </button>
        </div>
        <span
          class="MuiTypography-root MuiTypography-caption css-1374baf-MuiTypography-root"
        >
          Máximo 
          50
           unidades
        </span>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSecondary MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSecondary MuiButton-fullWidth css-15sow52-MuiButtonBase-root-MuiButton-root"
        id=":r5:"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeLarge css-hev6se-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
        </span>
        Already in Cart
      </button>
    </div>
  </div>
</div>
`;

exports[`ProductDetails should match snapshot when product is not in cart 1`] = `
<div>
  <div
    class="MuiBox-root css-19midj6"
  >
    <div
      data-testid="product-header"
    >
      Test Product
       - 
      4.5
       - 
      Test Category
    </div>
    <h4
      class="MuiTypography-root MuiTypography-h4 css-1uuypgi-MuiTypography-root"
    >
      $
      99.99
    </h4>
    <p
      class="MuiTypography-root MuiTypography-body1 css-o5kk20-MuiTypography-root"
    >
      Test description
    </p>
    <div
      class="MuiBox-root css-1yjvs5a"
    >
      <div
        data-testid="product-specs"
      >
        FBX,OBJ
         - 
        High
         - 
        4K
      </div>
    </div>
    <div
      class="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <div
        class="MuiBox-root css-j0ozid"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-k16oju-MuiTypography-root"
        >
          Cantidad:
        </p>
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-schvjd-MuiPaper-root"
        >
          <button
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeSmall css-1374xo4-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            id=":r0:"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiIconButton-loadingIndicator css-165vu30-MuiIconButton-loadingIndicator"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="RemoveIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13H5v-2h14z"
              />
            </svg>
          </button>
          <p
            class="MuiTypography-root MuiTypography-body1 css-144jna6-MuiTypography-root"
          >
            1
          </p>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1374xo4-MuiButtonBase-root-MuiIconButton-root"
            id=":r1:"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-loadingIndicator css-165vu30-MuiIconButton-loadingIndicator"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="AddIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
              />
            </svg>
          </button>
        </div>
        <span
          class="MuiTypography-root MuiTypography-caption css-1374baf-MuiTypography-root"
        >
          Máximo 
          50
           unidades
        </span>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-fullWidth css-1bbsvxk-MuiButtonBase-root-MuiButton-root"
        id=":r2:"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeLarge css-hev6se-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
        </span>
        Add to Cart
      </button>
    </div>
  </div>
</div>
`;
